RFC - feat: Introduce filtering by orders total on the Admin UI

adevinwild opened the pull request:
## Summary

**What** — What changes are introduced in this PR?

This PR introduce filtering by order total on the Admin UI and on the API side using operators ($eq, $gte, $lte...)

**Why** — Why are these changes relevant or necessary?  

Currently it is not possible

**How** — How have these changes been implemented?

To achieve this multiple parts of the code have been updated : 
- The `AdminGetOrdersParams` API validator had to be modified to accept the `total` key, which will then be mapped to the correct field on the `order_summary` table ( `total` is mapped to `summary.totals.current_order_total`, as it is the field synecd with the order's total)
- Fix of the operators insertion in the URL on the `NumberFilter` component as it was not matching the `createOperatorMap` validator util _( operator was not inserted with prefix `$` for the API, e.g `gt` instead of `$gt` )_
- Added new filters on the orders table filters

**Testing** — How have these changes been tested, or how can the reviewer test the feature?

- Build the packages
- Try to filter by Order total

---

## Examples

You can filter orders using the API by doing so : 
```
http://localhost:9000/admin/orders?total[<operator>]=<value>
```

For example if you want to fetch all orders with a total of $100 :
```
http://localhost:9000/admin/orders?total[$eq]=100
```
If you want greater than $100 : 
```
http://localhost:9000/admin/orders?total[$gt]=100
```

You get the idea, you can use all the common operators that are used in Medusa and listed here :
https://docs.medusajs.com/resources/service-factory-reference/tips/filtering#supported-filter-operators-list

---

## Checklist

Please ensure the following before requesting a review:

- [X] I have added a **changeset** for this PR
    - Every non-breaking change should be marked as a **patch**
    - To add a changeset, run `yarn changeset` and follow the prompts
- [ ] The changes are covered by relevant **tests**
- [X] I have verified the code works as intended locally
- [X] I have linked the related issue(s) if applicable

---

## Additional Context

Related to the following RFC 
https://github.com/medusajs/medusa/discussions/14154

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> **Medium Risk**
> Extends the admin orders list API validator and transforms query params into nested `summary` filters, which could affect order-list querying and performance if mis-mapped. UI query/filter serialization changes could also break existing saved views or URLs if consumers relied on the previous non-`$` operator format.
> 
> **Overview**
> Adds **order total filtering** end-to-end: the Admin API now accepts a `total` filter with operator maps and transforms it to `summary.totals.current_order_total` for querying.
> 
> Updates the Admin dashboard orders list to expose a *Total* number filter, plumbs `total` through the order table query parsing, and fixes number filter serialization to use `$`-prefixed operators (e.g. `$gt`, `$lt`, `$eq`) to match backend operator parsing.
> 
> Includes compatibility tweaks for table filter identifiers (mapping legacy `key` to `id` and guarding JSON parsing) and adds integration coverage for `GET /admin/orders` total filtering with operators.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit fc00c615903488a7134aeee17581d380433ce03f. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

adevinwild committed:
fix(number-filter): update operator key format in selectedParams for consistency

Based on the `createOperatorMap`, changed the operator key in selectedParams from `[operator]` to `[$${operator}]` to ensure consistent formatting across the filter logic (e.g `$lte` instead of `lte`)

adevinwild committed:
feat: added `total` in the `AdminGetOrdersParams`
- this points to the order_summary table

adevinwild committed:
feat(admin/dashboard): add `total` filter to order table query and filters

- Introduced a new filter for `total` in the order table query and updated the filters to include it.
- This enhancement allows users to filter orders based on the total amount.

adevinwild committed:
chore: version bump

adevinwild committed:
chore: version bump (patch for @medusajs/medusa instead of minor)

adevinwild committed:
fix(number-filter): update operator keys to directly include `$` prefix

adevinwild commented on packages/admin/dashboard/src/components/table/data-table/data-table-filter/number-filter.tsx:
```diff
@@ -60,17 +60,17 @@ export const NumberFilter = ({
         }
 
-        selectedParams.add(JSON.stringify({ ...curr, [operator]: value }))
+        selectedParams.add(JSON.stringify({ ...curr, [`$${operator}`]: value }))
```
Indeed, fixed at the operator type + fn level

adrien2p reviewed (dismissed):
This looks good to me, do you have a video or screen shot of the result?

adevinwild committed:
fix(number-filter): correct operator key references to include `$` prefix in parsing logic

adevinwild committed:
fix(order-repository): enhance summary filtering by adding support for explicit versioning

so far if we just try to filter summary total, it will get all versions, leading to potential inconsistency if you have multiple versions for an order when you want to filter by total

adevinwild committed:
Revert "fix(order-repository): enhance summary filtering by adding support for explicit versioning"

This reverts commit 483dac7fd3c460c886e636d1e79dfc4c32d536de.

adevinwild committed:
fix(configurable-data-table): ensure filter compatibility by mapping keys to IDs

Updated the filter processing logic to assign 'key' as 'id' when 'id' is not present, enhancing compatibility with existing filter structures. This change ensures that filters are correctly parsed and utilized in query parameters.

adrien2p commented on packages/medusa/src/api/admin/orders/validators.ts:
```diff
@@ -9,15 +9,12 @@ import {
-        z.number().optional()
-    )
+    version: z.preprocess((val) => {
+      if (val && typeof val === "string") {
```
suggestion: could we use `isString`

adrien2p reviewed (approved):
thanks man 🎉 should we add a new integration tests that filters with summary total on the order?

adevinwild committed:
refactor(order-validators): update version preprocessing to use isString utility for improved type checking

adevinwild committed:
test(order.spec): add tests for filtering orders by total using $eq, $gte, and $lte operators

Implemented a series of tests to verify the functionality of filtering orders based on total using various operators. The tests cover matching and non-matching scenarios for each operator, ensuring accurate filtering behavior.

adrien2p commented:
> > thanks man 🎉 should we add a new integration tests that filters with summary total on the order?
> 
> hey @adrien2p , I addressed the feedacks, is it ok to have a `describe` in the tests this way to better organize the operators ? 🤔

Yes thats alright, If I could make a suggestion it would be to maybe use `it.each` with operator and object for filter and expectation. So that if new operators would be added in the future it would make it easier to add new test expectation for new operators. But no strong opinion on that let me know what you think

adevinwild committed:
test(order.spec): refactor order filtering tests to use it.each

adevinwild committed:
Merge branch 'develop' into add-filtering-by-order-total

adevinwild committed:
fix: ensure valid version parsing in order parameters validator

NicolasGorga commented on packages/admin/dashboard/src/components/table/configurable-data-table/configurable-data-table.tsx:
```diff
@@ -76,9 +83,9 @@ export function ConfigurableDataTable<TData>({
-    const filterKey = filter.id
-    if (parsedQueryParams[filterKey] !== undefined) {
+  filters.forEach((filter) => {
+    const filterKey = (filter as any).id || (filter as any).key
```
nit: this filterKey could be `filter.id` as it is guaranteed to hold the value of id or key (if any present) as per the check in line 53 no?

adevinwild replied:
> nit: this filterKey could be `filter.id` as it is guaranteed to hold the value of id or key (if any present) as per the check in line 53 no?

Make sense, thanks for the heads up 🧠

NicolasGorga reviewed (approved):
lgtm

NicolasGorga committed:
Merge branch 'develop' into add-filtering-by-order-total

adevinwild committed:
fix: use of `filter.id` directly

adevinwild committed:
fix missing brackets

fPolic commented on .changeset/brave-forks-dress.md:
```diff
@@ -0,0 +1,6 @@
+---
+"@medusajs/dashboard": minor
```
q: why does dashboard has a minor patch

adevinwild replied:
> q: why does dashboard has a minor patch

I thought it would be more correct since it's not a fix, but I can make it a patch instead if that makes sense in terms of versioning. wdyt? 🙌

NicolasGorga replied:
Let's make it a patch @adevinwild Also if you could change the changeset message to `feat(medusa,dashboard): Introduce filtering orders by total price on the API and Admin UI`. Lastly, there are some conflicts if you could resolve

fPolic reviewed (commented):
Think it looks good, can we resolve the conflict and I'll do another pass

adevinwild committed:
merge: resolve conflict with upstream, migrate total filter to adapter

adevinwild committed:
merge: upstream/develop, keep total filter tests
